[jit][tracer] allow traced modules to return dicts with tuple values when strict=False

bradleyhd opened the pull request:
Summary: We have some inference use cases where the expected output of a module is of the form `{"key": (t1, t1)}` and are currently jit tracing the modules until we can reach jit script compatibility.

Test Plan: buck test mode/dev caffe2/test:jit -- 'test_trace_returning_complex_dict'

Differential Revision: D25624152

houseroad commented on test/jit/test_tracer.py:
```diff
@@ -2279,3 +2279,27 @@ def forward(self, input: Dict[str, Tuple[torch.Tensor, torch.Tensor, torch.Tenso
+                super().__init__()
+
+            def forward(
+                self, t: torch.Tensor
```
have two inputs, one for key, one for value? sometimes, the constant just hides the problem.

houseroad reviewed (approved):
Looks good.

wanchaol commented on torch/csrc/jit/frontend/tracer.cpp:
```diff
@@ -282,13 +282,15 @@ Value* TracingState::getOutput(const IValue& iv, size_t i) {
         key_type->isSubtypeOf(TensorType::get());
-    bool value_type_valid = value_type->isSubtypeOf(TensorType::get());
+    bool value_type_valid = value_type->isSubtypeOf(TensorType::get()) ||
+        value_type->isSubtypeOf(AnyTupleType::get());
```
This means the value type can be any of tuple type, it might actually including something like `Tuple[int, int]`? In that case it won't be a valid input right? The valid type check here need to be more strict.

houseroad replied:
Yeah, let's be strict on this check, we probably don't want to enable cases like Dict[str, Tuple[int, int]]

bradleyhd replied:
sounds good. updated by checking the contained types of the tuple, and added a test condition. let me know!

houseroad replied:
thanks, looks good to me.

@wanchaol, any more comment?

wanchaol reviewed (approved):
looks good, thanks!

bradleyhd committed:
[jit][tracer] allow traced modules to return dicts with tuple values when strict=False (#49568)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/49568

We have some inference use cases where the expected output of a module is of the form `{"key": (t1, t1)}` and are currently jit tracing the modules until we can reach jit script compatibility.

Test Plan: buck test mode/dev caffe2/test:jit -- 'test_trace_returning_complex_dict'

Reviewed By: houseroad

Differential Revision: D25624152

fbshipit-source-id: a2cf206ac377c93e1110ddc6827c1112fd6d4f45
